umich_api/api_utils.py

```diff
@@ -114,21 +125,44 @@ def _api_call(self, api_call, access_token, method="GET", payload=None):
             return resp
 
     def get_next_page(self, response):
+        #type: (requests.Response) -> Dict
         if not 'next' in response.links:
             return None
         self.__log.debug(response.links)
         return parse_qs(urlparse(response.links['next']['url']).query)
 
     def get_access_token(self, token_url, client_scope):
-        # type: (str) -> str
-        """Retrieves an access token from the specified URL
-        
+        # type: (str, str) -> str
+        """Retrieves access token from the local tokens dictionary
+           If it's expired it will attempt to renew it
         :param token_url: Full URL to retrieve access token
         :type token_url: str
         :param client_scope: Client scope to retrieve this token for
         :type client_scope: str
+        :returns Access token str
         """
+        
+        return self.get_oauth_token(token_url, client_scope).get("access_token")
 
+    def get_oauth_token(self, token_url, client_scope):
+        # type: (str, str) -> Dict
+        """Retrieves an access token from the specified URL, also caches the token in self.tokens dict
+        :param token_url: Full URL to retrieve access token
+        :type token_url: str
+        :param client_scope: Client scope to retrieve this token for
+        :type client_scope: str
+        :returns Dict (from json) generated from the response containing token information
+        """
+        if client_scope in self.tokens:
+            cached_token = self.tokens.get(client_scope)
+            # Check expires_time, if this is valid just return the token other-wise renew it, have a fallback incase this cam't be looked up
+            if (datetime.now() < cached_token.get("expires_time", datetime.min)):
```

Comment:
This is good that you created `expires_time`to check if expiration is nearing. But isn't it more appropriate to check the time diff between the `now() - expire_time() = 1sec` you renew the token than waiting for it last minute. 
```
if( now -expire_time == 1 sec)
     "renew token"
else
    "cached token"
 this way the content API call next will always ensure data

Reply:
Yeah I guess it should have a little bit of "padding".

Otherwise it's possible the time it checks the token to the time it gets to the call it could be expired. This is a good idea.

Reply:
Okay, I added a new named parameter token_expires_percent (default 5) which will give 3 minutes in a 1 hour token till expiration. This is fixed when the expiration time is set rather than calculated.